src/coreclr/src/gc/gc.cpp

```diff
@@ -8276,9 +8352,142 @@ void gc_heap::sort_mark_list()
+    size_t total_ephemeral_size = 0;
+    uint8_t* low = (uint8_t*)~0;
+    uint8_t* high = 0;
+    for (int i = 0; i < n_heaps; i++)
+    {
+        gc_heap* hp = g_heaps[i];
+        size_t ephemeral_size = heap_segment_allocated (hp->ephemeral_heap_segment) - hp->gc_low;
+        total_ephemeral_size += ephemeral_size;
+        total_mark_list_size += (hp->mark_list_index - hp->mark_list);
+        low = min (low, hp->gc_low);
+        high = max (high, heap_segment_allocated (hp->ephemeral_heap_segment));
+    }
+
+    // give up if this is not an ephemeral GC or the mark list size is unreasonably large
+    if (total_mark_list_size > total_ephemeral_size/256)
+    {
+        mark_list_index = mark_list_end + 1;
+        // let's not count this as a mark list overflow
+        mark_list_overflow = false;
+        return;
+    }
+
+#ifdef USE_VXSORT
+    ptrdiff_t item_count = mark_list_index - mark_list;
+    // conservatively use AVX2 only for large mark lists,
+    // and do runtime test to check whether AVX2 is indeed available 
+    if (item_count > AVX2_THRESHOLD_SIZE && IsSupportedInstructionSet (InstructionSet::AVX2))
+    {
+#if defined(_DEBUG) || defined(WRITE_SORT_DATA)
+        // in debug, make a copy of the mark list
+        // for checking and debugging purposes
+        uint8_t** mark_list_copy = &g_mark_list_copy[heap_number * mark_list_size];
+        uint8_t** mark_list_copy_index = &mark_list_copy[item_count];
+        for (ptrdiff_t i = 0; i < item_count; i++)
+        {
+            uint8_t* item = mark_list[i];
+            mark_list_copy[i] = item;
+        }
+#endif // defined(_DEBUG) || defined(WRITE_SORT_DATA)
+
+        // is the range small enough for a 32-bit sort?
+        ptrdiff_t range = high - low;
+        assert(sizeof(uint8_t*) == (1<<3));
+        ptrdiff_t scaled_range = range >> 3;
+        if ((uint32_t)scaled_range == scaled_range)
+        {
+            dprintf (3, ("Sorting mark lists as 32-bit offsets"));
+
+//#define WRITE_SORT_DATA
+
+            // first step: scale the pointers down to 32-bit offsets
+            uint8_t** mark_list = this->mark_list;
+            int32_t* mark_list_32 = (int32_t*)mark_list;
+            do_pack_avx2 (mark_list, item_count, low);
+            // sort the 32-bit offsets
+            if (item_count > 0)
+            {
+                ptrdiff_t start = get_cycle_count();
+
+                _sort (&mark_list_32[0], &mark_list_32[item_count - 1], 0);
+
+                ptrdiff_t elapsed_cycles = get_cycle_count() - start;
+                int log2_item_count = index_of_highest_set_bit (item_count);
+                double elapsed_cyles_by_n_log_n = (double)elapsed_cycles / item_count / log2_item_count;
+
+//                printf ("GC#%d: first phase of sort_mark_list for heap %d took %u cycles to sort %u entries (cost/(n*log2(n) = %5.2f)\n", settings.gc_index, this->heap_number, elapsed_cycles, item_count, elapsed_cyles_by_n_log_n);
```

Comment:
are you actually using the printf's in this method? if not we should just get rid of them... or convert to dprintf's...

Reply:
I converted the printfs to dprintfs.

The call to introsort is a debug only check. We can get rid of it once we fully trust the vectorized sort.